modules/rostests/kmtests/ntos_io/IoDeviceInterface.c

```diff
@@ -1,125 +1,214 @@
-
-   Status = IoGetDeviceInterfaces(
-      &Guid,
-      NULL,
-      0,
-      &SymbolicLinkList);
-
-   ok(NT_SUCCESS(Status),
-         "IoGetDeviceInterfaces failed with status 0x%X\n",
-         (unsigned int)Status);
-   if (!NT_SUCCESS(Status))
-   {
-      return;
-   }
-
-   DPRINT("IoGetDeviceInterfaces results:\n");
-   for (SymbolicLinkListPtr = SymbolicLinkList;
-        SymbolicLinkListPtr[0] != 0 && SymbolicLinkListPtr[1] != 0;
-        SymbolicLinkListPtr += wcslen(SymbolicLinkListPtr) + 1)
-   {
-      DPRINT1("Symbolic Link: %S\n", SymbolicLinkListPtr);
-   }
+    &KSCATEGORY_BRIDGE,
+    &KSCATEGORY_CAPTURE,
+    &KSCATEGORY_COMMUNICATIONSTRANSFORM,
+    &KSCATEGORY_DATACOMPRESSOR,
+    &KSCATEGORY_DATADECOMPRESSOR,
+    &KSCATEGORY_DATATRANSFORM,
+    &KSCATEGORY_FILESYSTEM,
+    &KSCATEGORY_INTERFACETRANSFORM,
+    &KSCATEGORY_MEDIUMTRANSFORM,
+    &KSCATEGORY_MIXER,
+    &KSCATEGORY_RENDER,
+    &KSCATEGORY_SPLITTER,
+};
 
-#if 0
-   DPRINT("[PnP Test] Trying to get aliases\n");
+static
+VOID
+Test_IoOpenDeviceInterfaceRegistryKey(PWSTR SymbolicLink)
+{
+    UNICODE_STRING KeyName, SymbolicLinkName;
 
-   for (SymbolicLinkListPtr = SymbolicLinkList;
-        SymbolicLinkListPtr[0] != 0 && SymbolicLinkListPtr[1] != 0;
-        SymbolicLinkListPtr += wcslen(SymbolicLinkListPtr) + 1)
-   {
-      UNICODE_STRING SymbolicLink;
-      UNICODE_STRING AliasSymbolicLink;
-
-      SymbolicLink.Buffer = SymbolicLinkListPtr;
-      SymbolicLink.Length = SymbolicLink.MaximumLength = wcslen(SymbolicLinkListPtr);
-      RtlInitUnicodeString(&AliasSymbolicLink, NULL);
-      IoGetDeviceInterfaceAlias(
-         &SymbolicLink,
-         &AliasGuid,
-         &AliasSymbolicLink);
-      if (AliasSymbolicLink.Buffer != NULL)
-      {
-         DPRINT("[PnP Test] Original: %S\n", SymbolicLinkListPtr);
-         DPRINT("[PnP Test] Alias: %S\n", AliasSymbolicLink.Buffer);
-      }
-   }
-#endif
-
-   ExFreePool(SymbolicLinkList);
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+    RtlInitUnicodeString(&KeyName, L"ReactOS_kmtest");
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        HANDLE DeviceInterfaceKey, DeviceInterfaceSubKey;
+        OBJECT_ATTRIBUTES ObjectAttributes;
+
+        NTSTATUS Status = IoOpenDeviceInterfaceRegistryKey(&SymbolicLinkName, KEY_CREATE_SUB_KEY, &DeviceInterfaceKey);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoOpenDeviceInterfaceRegistryKey(): success: %d %p\n", n, DeviceInterfaceKey);
+        }
+        else
+        {
+            trace("IoOpenDeviceInterfaceRegistryKey(): fail: %d 0x%x\n", n, Status);
+        }
+
+        InitializeObjectAttributes(&ObjectAttributes,
+                                   &KeyName,
+                                   OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                                   DeviceInterfaceKey,
+                                   NULL);
+        Status = ZwCreateKey(&DeviceInterfaceSubKey,
```

Comment:
What's the purpose of creating a subkey? If you just want to validate that the handle is valid, `ObReferenceObjectByHandle` with the handle and `CmKeyObjectType` will do the job. Or otherwise some read-only operation, like reading the default value or enumerating values/subkeys.

Reply:
The purpose of the subkey is to check the create options of the parent key: volatile or non-volatile. It was required for creating a proper fix for #3583 PR. If the first call didn't fail, the handle obviously seems to be valid I think. At least in case with `IoOpenDeviceInterfaceRegistryKey`, as said in MS documentation.

Reply:
Gotcha. You're right, looks like creating a nonvolatile subkey is the only way to verify whether a key is volatile.
Just remember to use `ok()` to actually test something, then.

Reply:
OK, added a comment and an `ok()` call.